### ISSUE TITLE
Validating saving checkpoints (peculiar issue with reloading)

### ISSUE THREAD

**Issue by geraltFromRivia** [cdluminate/robrank#18]
Hi Mo,

What command do you use to obtain the metrics (R@1, mAP, NMI) on benign examples?

I've used your code to train a robust retrieval model (cub-rres18ghmetsm-ptripletN) for the CUB dataset, and at the end I get validation accuracy similar to the one you reported in your CVPR '22 paper (taken from logs):

    Validation │ r@M= 18.3 r@1= 0.376 r@2= 0.485 mAP= 0.172 mAP@R = 0.222 NMI= 0.485

The problem arises, when I try to reload the checkpoint through either validate or continue training, I get very different results:

    Validation │ r@M= 6.8 r@1= 0.450 r@2= 0.584 mAP= 0.164 mAP@R = 0.239 NMI= 0.546

The problem is that I get very similar "weird" results no matter what checkpoint I pass into the Train / Validate Class in cmdline.py.

Thanks!

**Comment by geraltFromRivia** [comment]
Further question: in your paper, do you report mAP or mAP@R in the tables?

**Comment by cdluminate** [comment]
> What command do you use to obtain the metrics (R@1, mAP, NMI) on benign examples?
> 
> I've used your code to train a robust retrieval model (cub-rres18ghmetsm-ptripletN) for the CUB dataset, and at the end I get validation accuracy similar to the one you reported in your CVPR '22 paper (taken from logs):
> 
> ```
> Validation │ r@M= 18.3 r@1= 0.376 r@2= 0.485 mAP= 0.172 mAP@R = 0.222 NMI= 0.485
> ```
> 
> The problem arises, when I try to reload the checkpoint through either validate or continue training, I get very different results:
> 
> ```
> Validation │ r@M= 6.8 r@1= 0.450 r@2= 0.584 mAP= 0.164 mAP@R = 0.239 NMI= 0.546
> ```
> 
> The problem is that I get very similar "weird" results no matter what checkpoint I pass into the Train / Validate Class in cmdline.py.

Is that the first validation result after you reload the model? This is a pytorch-lightning feature, where it runs a fake validation epoch (with merely several batches instead of traversing through the whole validation set) before the training process, for debugging purposes (I guess).


The `tools/pjswipe.py` script is provided to dump the evaluation results in a nice format. The evaluation results are stored in the JSON file just beside the checkpoint file you specified.

I added some more documentation on the complete command line pipeline for reproducing the work. See https://github.com/cdluminate/robrank#14-scripts-for-complete-pipeline

**Comment by cdluminate** [comment]
> Further question: in your paper, do you report mAP or mAP@R in the tables?

What was reported was just mAP.
mAP@R is just implemented and reported for reference.

**Comment by cdluminate** [comment]
Let me know if your concerns are resolved.

**Comment by geraltFromRivia** [comment]
> > What command do you use to obtain the metrics (R@1, mAP, NMI) on benign examples?
> > I've used your code to train a robust retrieval model (cub-rres18ghmetsm-ptripletN) for the CUB dataset, and at the end I get validation accuracy similar to the one you reported in your CVPR '22 paper (taken from logs):
> > ```
> > Validation │ r@M= 18.3 r@1= 0.376 r@2= 0.485 mAP= 0.172 mAP@R = 0.222 NMI= 0.485
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The problem arises, when I try to reload the checkpoint through either validate or continue training, I get very different results:
> > ```
> > Validation │ r@M= 6.8 r@1= 0.450 r@2= 0.584 mAP= 0.164 mAP@R = 0.239 NMI= 0.546
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The problem is that I get very similar "weird" results no matter what checkpoint I pass into the Train / Validate Class in cmdline.py.
> 
> Is that the first validation result after you reload the model? This is a pytorch-lightning feature, where it runs a fake validation epoch (with merely several batches instead of traversing through the whole validation set) before the training process, for debugging purposes (I guess).
> 
> The `tools/pjswipe.py` script is provided to dump the evaluation results in a nice format. The evaluation results are stored in the JSON file just beside the checkpoint file you specified.
> 
> I added some more documentation on the complete command line pipeline for reproducing the work. See https://github.com/cdluminate/robrank#14-scripts-for-complete-pipeline

Thank you for the fast reply and for posting the script!

That is the first result after reloading the model, however I ran "trainer.validate(model)".

I am aware of the lightning doing a sanity check, adn therefore I changed "num_sanity_val_steps" to

    num_sanity_val_steps=-1.

I use 

    python bin/swipe.py -p rob224 -C ${CKPTPATH}

to calculate the evaluate the model against adversarial attacks, however tools/pjswipe.py only gives you values for CA+, etc. Do you extract R@1, R@2, mAP, and NMI for benign examples from the training logs?

**Comment by cdluminate** [comment]
> 
> I use
> 
> ```
> python bin/swipe.py -p rob224 -C ${CKPTPATH}
> ```
> 
> to calculate the evaluate the model against adversarial attacks, however tools/pjswipe.py only gives you values for CA+, etc. Do you extract R@1, R@2, mAP, and NMI for benign examples from the training logs?

Install tensorflow-cpu (or tensorflow-gpu) and run pjswipe.py again. If you read the screen outputs, you will also find
```
tensorflow-cpu not installed cannot dump tensorboard binary
```

**Comment by geraltFromRivia** [comment]
> > I use
> > ```
> > python bin/swipe.py -p rob224 -C ${CKPTPATH}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to calculate the evaluate the model against adversarial attacks, however tools/pjswipe.py only gives you values for CA+, etc. Do you extract R@1, R@2, mAP, and NMI for benign examples from the training logs?
> 
> Install tensorflow-cpu (or tensorflow-gpu) and run pjswipe.py again. If you read the screen outputs, you will also find
> 
> ```
> tensorflow-cpu not installed cannot dump tensorboard binary
> ```

Of course! I will install it right away. I think I found a temporary solution to model evaluations. I tried to load the model weights with 

    model.load_from_checkpoint(ag.checkpoint, dataset=ag.dataset, loss=ag.loss)

or

     resume_from_checkpoint=ag.checkpoint in Trainer

however neither seemed to work.

What works for me is to load the state dictionary directly into the model,

    checkpoint = th.load( ag.checkpoint )
    model.load_state_dict( checkpoint['state_dict'] )

after which the evaluation loop matches the one in the logs.

Thank you for all of the help!
